demos/tcp/aws_tcp_echo_client_single_task.c

```diff
@@ -350,6 +350,11 @@ static void prvEchoClientTask( void * pvParameters )
                 }
                 else
                 {
+                    /* The received length did not match the transmitted
+                     * length. */
+                    ulTxRxFailures[ xInstance ]++;
```

Comment:
It seems like the handling for this condition (xTransmitted>xReceivedBytes>=0) might be swapped with the case above (xReceivedBytes < 0)

I think we should trigger an error condition on xReceivedBytes being negative because that indicates an error occurred.

I'm not as clear on whether  (xTransmitted>xReceivedBytes>=0) should simply break and try again, or if we consider not receiving all the bytes within the timeout an error and we should increment the failure counter.

Reply:
I'd also like to see negative values treated separately and brought to the user's attention.  (Do we have a perror(3) equivalent?)

Reply:
The error numbers returned by the API's are quite compatible with errno.h
I sometimes use `strerror_r()` to print them:

    if( xReceivedBytes < 0 )
    {
        char pcBuffer[ 128 ];
        strerror_r( -xReceivedBytes, pcBuffer, sizeof( pcBuffer ) );
        configPRINTF( ( "ERROR: xTransmitted %d errno %d (%s)\r\n",
                         int ) xTransmitted, ( int ) -xReceivedBytes, pcBuffer ) );
    }
    else
    {
        configPRINTF( ( "ERROR: xTransmitted %d xReceivedBytes %d \r\n",
                         int ) xTransmitted, ( int ) xReceivedBytes ) );
    }
    break;

But I'm afraid that `strerror_r()` is too compiler-specific?

Reply:
In the mqtt demo I see IotMqtt_strerror() where someone rolled their own.  We could do something like that with a small subset of meaningful errors.

Reply:
> In the mqtt demo I see IotMqtt_strerror()

FreeRTOS+TCP uses 15 different errno values, so it is doable to make a `strerror()` function.
```c
EADDRINUSE          ENOMEM
EADDRNOTAVAIL       ENOPROTOOPT
EBADF               ENOSPC
ECANCELED           ENOTCONN
EINPROGRESS         EOPNOTSUPP
EINTR               ETIMEDOUT
EINVAL              EWOULDBLOCK /* same as EAGAIN */
EISCONN

```
I don't mind making a separate PR to introduce that function.